src/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -66,38 +47,223 @@ internal static unsafe IPAddress Parse(string ipString, bool tryParse)
+        {
+            Debug.Assert(ipString != null);
+            Debug.Assert(numbers != null);
+            Debug.Assert(numbersLength >= IPAddressParserStatics.IPv6AddressShorts);
+
+            int end = ipString.Length;
+            fixed (char* name = ipString)
+            {
+                if (IPv6AddressHelper.IsValidStrict(name, 0, ref end) || (end != ipString.Length))
+                {
+                    string scopeId = null;
+                    IPv6AddressHelper.Parse(ipString, numbers, 0, ref scopeId);
+
+                    long result = 0;
+                    if (!string.IsNullOrEmpty(scopeId))
+                    {
+                        if (scopeId.Length < 2)
+                        {
+                            scope = 0;
+                            return false;
+                        }
+
+                        for (int i = 1; i < scopeId.Length; i++)
+                        {
+                            char c = scopeId[i];
+                            if (c < '0' || c > '9')
+                            {
+                                scope = 0;
+                                return false;
+                            }
+                            result = (result * 10) + (c - '0');
+                            if (result > uint.MaxValue)
+                            {
+                                scope = 0;
+                                return false;
+                            }
+                        }
+                    }
+
+                    scope = (uint)result;
+                    return true;
+                }
+            }
+
+            scope = 0;
+            return false;
+        }
+
+        /// <summary>
+        /// Appends each of the numbers in address in indexed range [fromInclusive, toExclusive),
+        /// while also replacing the longest sequence of 0s found in that range with "::", as long
+        /// as the sequence is more than one 0.
+        /// </summary>
+        private static void AppendSections(ushort[] address, int fromInclusive, int toExclusive, StringBuilder buffer)
+        {
+            // Find the longest sequence of zeros to be combined into a "::"
+            (int zeroStart, int zeroEnd) = IPv6AddressHelper.FindCompressionRange(address, fromInclusive, toExclusive);
+            bool needsColon = false;
+
+            // Output all of the numbers before the zero sequence
+            for (int i = fromInclusive; i < zeroStart; i++)
+            {
+                if (needsColon)
+                {
+                    buffer.Append(':');
+                }
+                needsColon = true;
+                AppendHex(address[i], buffer);
+            }
+
+            // Output the zero sequence if there is one
+            if (zeroStart >= 0)
+            {
+                buffer.Append("::");
+                needsColon = false;
+                fromInclusive = zeroEnd;
```

Comment:
It's a bit odd that we assign fromInclusive here.  Why not just use zeroEnd below and avoid mutating these, for clarity?